File: typescript/docs/snippets/session.server.ts
// Server-side session: per-chunk billing with off-chain vouchers + on-chain settle at close.
// See ./README.md for the snippet:start/end convention.

import express from 'express'
import type { KeyPairSigner } from '@solana/kit'
import { createPayKit, session, usd } from '@solana/pay-kit'

declare const signer: KeyPairSigner
declare const RECIPIENT: string
declare const rpcUrl: string
declare function chunks(): Iterable<string>

// snippet:start
const pay = await createPayKit({
  network: 'mainnet',
  operator: { recipient: RECIPIENT, signer },
  // Cap 1 USDC per session, metered at 0.0001 USDC per delivered chunk.
  pricing: { stream: session(usd('1.00'), { unitPrice: usd('0.0001') }) },
  rpcUrl,
})

const app = express()

// Gate the stream; settlement runs out-of-band when the channel idle-closes.
app.get('${PATH}', pay.express('stream'), (_req, res) => {
  res.writeHead(200, { 'Content-Type': 'text/event-stream' })
  for (const chunk of chunks()) res.write(`data: ${chunk}\n\n`)
  res.end()
})

// Session side-channel + receipt routes — mounted explicitly (pay-kit, like
// mppx, leaves route-mounting to the app rather than auto-injecting them).
const routes = pay.sessionRoutes('stream')
app.post('${PATH}', routes.voucher) // voucher commits to the resource URL
app.post('/__402/session/deliveries', routes.deliveries)
app.post('/__402/session/commit', routes.commit)
app.get('/sessions/receipt/:channelId', routes.receipt)
// snippet:end

void app

File: python/docs/snippets/session.server.py
# Server-side session: a metered payment channel, billed per delivery.
#
# Mirrors examples/playground_api/sessions.py. See
# ../../../docs/snippets-convention.md for the snippet:start/end convention.
import solana_pay_kit
from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse, StreamingResponse
from solana_pay_kit._paycore.rpc import SolanaRpc
from solana_pay_kit._paycore.solana import resolve_mint, stablecoin_decimals
from solana_pay_kit.fastapi import RequireSession
from solana_pay_kit.protocols.mpp.server import (
    SessionChallengeOptions,
    SessionOptions,
    new_session,
    session_routes,
)

router = APIRouter()
cfg = solana_pay_kit.config()

# snippet:start
# One session method, built from the shared config. `cap` is the ceiling the
# server offers; on-chain settle-at-close needs the operator signer + RPC.
session = new_session(
    SessionOptions(
        operator=cfg.operator.signer.pubkey(),
        recipient=cfg.effective_recipient(),
        cap=1_000_000,  # 1.00 USDC ceiling
        currency=resolve_mint("USDC", "mainnet"),
        decimals=stablecoin_decimals("USDC"),
        network=cfg.network.value,
        secret_key=cfg.mpp.challenge_binding_secret or "",
        signer=cfg.operator.signer,
        rpc=SolanaRpc(cfg.effective_rpc_url()),
        open_tx_submitter="server",
        close_delay=2.0,  # idle-close watchdog
    )
)
challenge = SessionChallengeOptions(cap="1000000", description="Metered token stream")

# RequireSession is the 402 gate: verify a credential/voucher or answer with a
# fresh challenge — the session counterpart of RequirePayment.
gate = Depends(RequireSession(session, challenge))


@router.get("${PATH}")
async def stream(_=gate) -> StreamingResponse:
    ...  # stream metered deliveries, billed per chunk against the voucher


# Mount the reserve/commit metering side channel built by session_routes.
routes = session_routes(session.core())


@router.post("/__402/session/deliveries")
async def deliveries(request: Request) -> JSONResponse:
    r = await routes.deliveries(request.method, await request.body() or b"{}")
    return JSONResponse(r.body, status_code=r.status)
# snippet:end
